File: backend/src/main/java/com/lacrisalide/payment/PayPalAdapter.java
package com.lacrisalide.payment;

import com.lacrisalide.dto.donation.CapturePaymentResponse;
import com.lacrisalide.dto.donation.CreatePaymentRequest;
import com.lacrisalide.dto.donation.CreatePaymentResponse;
import com.lacrisalide.exception.BadRequestException;
import java.nio.charset.StandardCharsets;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Base64;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

@Component
@RequiredArgsConstructor
public class PayPalAdapter implements PaymentAdapter {

 private final RestTemplate restTemplate;

 @Value("${paypal.client-id:}")
 private String clientId;

 @Value("${paypal.client-secret:}")
 private String clientSecret;

 @Value("${paypal.base-url}")
 private String baseUrl;

 @Value("${paypal.currency:EUR}")
 private String currency;

 @Override
 public CreatePaymentResponse createPayment(CreatePaymentRequest request) {
  ensureConfigured();

  HttpHeaders headers = bearerHeaders(fetchAccessToken());
  headers.setContentType(MediaType.APPLICATION_JSON);

  Map<String, Object> payload = Map.of(
   "intent", "CAPTURE",
   "purchase_units", List.of(
    Map.of(
     "description", "Donazione La Crisalide - " + request.nome(),
     "amount", Map.of(
      "currency_code", currency,
      "value", formatAmount(request.importo())
     )
    )
   ),
   "payment_source", Map.of(
    "paypal", Map.of(
     "experience_context", Map.of(
      "payment_method_preference", "IMMEDIATE_PAYMENT_REQUIRED",
      "shipping_preference", "NO_SHIPPING",
      "user_action", "PAY_NOW"
     )
    )
   )
  );

  try {
   ResponseEntity<Map> response = restTemplate.exchange(
    baseUrl + "/v2/checkout/orders",
    HttpMethod.POST,
    new HttpEntity<>(payload, headers),
    Map.class
   );

   Map<String, Object> body = response.getBody();
   if (body == null) {
    throw new BadRequestException("PayPal non ha restituito una risposta valida");
   }

   return new CreatePaymentResponse(
    stringValue(body.get("id")),
    stringValue(body.get("status")),
    extractApprovalUrl(body)
   );
  } catch (RestClientException ex) {
   throw new BadRequestException("Errore durante la creazione dell'ordine PayPal");
  }
 }

 @Override
 public CapturePaymentResponse capturePayment(String orderId) {
  ensureConfigured();

  HttpHeaders headers = bearerHeaders(fetchAccessToken());
  headers.setContentType(MediaType.APPLICATION_JSON);

  try {
   ResponseEntity<Map> response = restTemplate.exchange(
    baseUrl + "/v2/checkout/orders/" + orderId + "/capture",
    HttpMethod.POST,
    new HttpEntity<>(headers),
    Map.class
   );

   Map<String, Object> body = response.getBody();
   if (body == null) {
    throw new BadRequestException("PayPal non ha restituito i dettagli della cattura");
   }

   Map<String, Object> payer = castMap(body.get("payer"));
   List<Map<String, Object>> purchaseUnits = castList(body.get("purchase_units"));
   Map<String, Object> payments = purchaseUnits.isEmpty() ? Map.of() : castMap(purchaseUnits.get(0).get("payments"));
   List<Map<String, Object>> captures = payments.isEmpty() ? List.of() : castList(payments.get("captures"));
   Map<String, Object> firstCapture = captures.isEmpty() ? Map.of() : captures.get(0);

   return new CapturePaymentResponse(
    stringValue(body.get("id")),
    stringValue(body.get("status")),
    stringValue(payer.get("payer_id")),
    stringValue(firstCapture.get("id"))
   );
  } catch (RestClientException ex) {
   throw new BadRequestException("Errore durante la cattura del pagamento PayPal");
  }
 }

 private String fetchAccessToken() {
  HttpHeaders headers = new HttpHeaders();
  headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
  headers.set(
   HttpHeaders.AUTHORIZATION,
   "Basic " + Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes(StandardCharsets.UTF_8))
  );

  MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
  body.add("grant_type", "client_credentials");

  try {
   ResponseEntity<Map> response = restTemplate.exchange(
    baseUrl + "/v1/oauth2/token",
    HttpMethod.POST,
    new HttpEntity<>(body, headers),
    Map.class
   );
   Map<String, Object> responseBody = response.getBody();
   if (responseBody == null || responseBody.get("access_token") == null) {
    throw new BadRequestException("Impossibile ottenere il token PayPal");
   }
   return responseBody.get("access_token").toString();
  } catch (RestClientException ex) {
   throw new BadRequestException("Autenticazione PayPal non riuscita");
  }
 }

 private HttpHeaders bearerHeaders(String accessToken) {
  HttpHeaders headers = new HttpHeaders();
  headers.setBearerAuth(accessToken);
  headers.setAccept(List.of(MediaType.APPLICATION_JSON));
  return headers;
 }

 private void ensureConfigured() {
  if (clientId == null || clientId.isBlank() || clientSecret == null || clientSecret.isBlank()) {
   throw new BadRequestException("Configura PAYPAL_CLIENT_ID e PAYPAL_CLIENT_SECRET per usare PayPal");
  }
 }

 @SuppressWarnings("unchecked")
 private Map<String, Object> castMap(Object value) {
  return value instanceof Map<?, ?> map ? (Map<String, Object>) map : Map.of();
 }

 @SuppressWarnings("unchecked")
 private List<Map<String, Object>> castList(Object value) {
  return value instanceof List<?> list ? (List<Map<String, Object>>) list : List.of();
 }

 private String extractApprovalUrl(Map<String, Object> body) {
  List<Map<String, Object>> links = castList(body.get("links"));
  return links.stream()
   .filter(link -> "approve".equalsIgnoreCase(stringValue(link.get("rel"))))
   .map(link -> stringValue(link.get("href")))
   .findFirst()
   .orElse(null);
 }

 private String formatAmount(Double amount) {
  DecimalFormat formatter = new DecimalFormat("0.00", DecimalFormatSymbols.getInstance(Locale.US));
  return formatter.format(amount);
 }

 private String stringValue(Object value) {
  return value == null ? null : value.toString();
 }
}

File: backend/Payments/PayPalAdapter.cs
using System.Globalization;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using SoccerDreamFermana.Backend.Data;
using SoccerDreamFermana.Backend.Dtos;
using SoccerDreamFermana.Backend.Exceptions;

namespace SoccerDreamFermana.Backend.Payments;

public sealed class PayPalAdapter(IHttpClientFactory httpClientFactory, IConfiguration configuration) : IPaymentAdapter
{
    private readonly string? _clientId = ConfigurationReader.Get(configuration, "PAYPAL_CLIENT_ID", "PayPal:ClientId");
    private readonly string? _clientSecret = ConfigurationReader.Get(configuration, "PAYPAL_CLIENT_SECRET", "PayPal:ClientSecret");
    private readonly string _baseUrl = ConfigurationReader.Get(configuration, "PAYPAL_BASE_URL", "PayPal:BaseUrl", "https://api-m.sandbox.paypal.com").TrimEnd('/');
    private readonly string _currency = ConfigurationReader.Get(configuration, "PAYPAL_CURRENCY", "PayPal:Currency", "EUR");

    public async Task<CreatePaymentResponse> CreatePaymentAsync(CreatePaymentRequest request, CancellationToken cancellationToken)
    {
        EnsureConfigured();

        var client = httpClientFactory.CreateClient();
        using var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v2/checkout/orders");
        httpRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", await FetchAccessTokenAsync(cancellationToken));
        httpRequest.Content = JsonContent.Create(new
        {
            intent = "CAPTURE",
            purchase_units = new[]
            {
                new
                {
                    description = "Donazione - Associazione Disabili " + request.Nome,
                    amount = new
                    {
                        currency_code = _currency,
                        value = request.Importo.ToString("0.00", CultureInfo.InvariantCulture)
                    }
                }
            },
            payment_source = new
            {
                paypal = new
                {
                    experience_context = new
                    {
                        payment_method_preference = "IMMEDIATE_PAYMENT_REQUIRED",
                        shipping_preference = "NO_SHIPPING",
                        user_action = "PAY_NOW"
                    }
                }
            }
        });

        using var response = await client.SendAsync(httpRequest, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new BadRequestException("Errore durante la creazione dell'ordine PayPal");
        }

        using var document = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(cancellationToken), cancellationToken: cancellationToken);
        var root = document.RootElement;
        return new CreatePaymentResponse(
            ReadString(root, "id"),
            ReadString(root, "status"),
            ExtractApprovalUrl(root)
        );
    }

    public async Task<CapturePaymentResponse> CapturePaymentAsync(string orderId, CancellationToken cancellationToken)
    {
        EnsureConfigured();

        var client = httpClientFactory.CreateClient();
        using var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v2/checkout/orders/{Uri.EscapeDataString(orderId)}/capture");
        httpRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", await FetchAccessTokenAsync(cancellationToken));
        httpRequest.Content = JsonContent.Create(new { });

        using var response = await client.SendAsync(httpRequest, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new BadRequestException("Errore durante la cattura del pagamento PayPal");
        }

        using var document = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(cancellationToken), cancellationToken: cancellationToken);
        var root = document.RootElement;
        var payerId = root.TryGetProperty("payer", out var payer) ? ReadString(payer, "payer_id") : null;
        string? captureId = null;

        if (root.TryGetProperty("purchase_units", out var purchaseUnits)
            && purchaseUnits.ValueKind == JsonValueKind.Array
            && purchaseUnits.GetArrayLength() > 0
            && purchaseUnits[0].TryGetProperty("payments", out var payments)
            && payments.TryGetProperty("captures", out var captures)
            && captures.ValueKind == JsonValueKind.Array
            && captures.GetArrayLength() > 0)
        {
            captureId = ReadString(captures[0], "id");
        }

        return new CapturePaymentResponse(ReadString(root, "id"), ReadString(root, "status"), payerId, captureId);
    }

    private async Task<string> FetchAccessTokenAsync(CancellationToken cancellationToken)
    {
        var client = httpClientFactory.CreateClient();
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v1/oauth2/token");
        request.Headers.Authorization = new AuthenticationHeaderValue(
            "Basic",
            Convert.ToBase64String(Encoding.UTF8.GetBytes($"{_clientId}:{_clientSecret}"))
        );
        request.Content = new FormUrlEncodedContent(new Dictionary<string, string> { ["grant_type"] = "client_credentials" });

        using var response = await client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new BadRequestException("Autenticazione PayPal non riuscita");
        }

        using var document = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(cancellationToken), cancellationToken: cancellationToken);
        return ReadString(document.RootElement, "access_token")
            ?? throw new BadRequestException("Impossibile ottenere il token PayPal");
    }

    private void EnsureConfigured()
    {
        if (string.IsNullOrWhiteSpace(_clientId) || string.IsNullOrWhiteSpace(_clientSecret))
        {
            throw new BadRequestException("Configura PAYPAL_CLIENT_ID e PAYPAL_CLIENT_SECRET per usare PayPal");
        }
    }

    private static string? ExtractApprovalUrl(JsonElement root)
    {
        if (!root.TryGetProperty("links", out var links) || links.ValueKind != JsonValueKind.Array)
        {
            return null;
        }

        foreach (var link in links.EnumerateArray())
        {
            if ("approve".Equals(ReadString(link, "rel"), StringComparison.OrdinalIgnoreCase))
            {
                return ReadString(link, "href");
            }
        }

        return null;
    }

    private static string? ReadString(JsonElement element, string property)
    {
        return element.TryGetProperty(property, out var value) && value.ValueKind != JsonValueKind.Null
            ? value.ToString()
            : null;
    }
}
